### PR TITLE
feat: use tracker shape size in detection by tracker

### PR DESCRIPTION
**Pull request by yukkysaito**
## Description
In [1], the author says `Bounding box area generally should not shrink`.
But, currently, the "detection by tracker" doesn't have the assumption.
In this PR, I implemented it so that the "detection by tracker" can refer the tracker shape size as a minimum value.

[1] Arya Senna Abdul Rachman, Arya. "3D-LIDAR Multi Object Tracking for Autonomous Driving: Multi-target Detection and Tracking under Urban Road Uncertainties." (2017).

https://youtu.be/VXdS4o7asBA

Not a big difference in the tutorial, but more stability for larger vehicles such as trucks

## Tests performed
Tutorial rosbag.
https://youtu.be/VXdS4o7asBA

Not a big difference in the tutorial, but more stability for larger vehicles such as trucks

## Pre-review checklist for the PR author

The PR author **must** check the checkboxes below when creating the PR.

- [x] I've confirmed the [contribution guidelines].
- [x] The PR follows the [pull request guidelines].

## In-review checklist for the PR reviewers

The PR reviewers **must** check the checkboxes below before approval.

- [ ] The PR follows the [pull request guidelines].
- [ ] The PR has been properly tested.
- [ ] The PR has been reviewed by the code owners.

## Post-review checklist for the PR author

The PR author **must** check the checkboxes below before merging.

- [ ] There are no open discussions or they are tracked via tickets.
- [ ] The PR is ready for merge.

After all checkboxes are checked, anyone who has write access can merge the PR.

[contribution guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/
[pull request guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/pull-request-guidelines/

### CODE REVIEW

**Comment by yukke42** [comment]
In this PR, detection_by_tracker and object_merger have changed at the same time, but there are not any description about it. Is this change what you intended?

**Comment by yukkysaito** [comment]
>In this PR, detection_by_tracker and object_merger have changed at the same time, but there are not any description about it. Is this change what you intended?

@yukke42 
Sorry for the lack of explanation.
Modes are added to the object merger to prioritize `detection by tracker` results.
- Confidence based (Original implement)
- Input based (Added in this PR)

**Review by yukke42 — COMMENTED** [review]
I have left some comments.

File: launch/tier4_perception_launch/launch/object_recognition/detection/lidar_based_detection.launch.xml [hunk 1]
```diff
@@ -125,6 +126,7 @@
       <arg name="input/object0" value="temporary_merged_objects"/>
       <arg name="input/object1" value="detection_by_tracker/objects"/>
       <arg name="output/object" value="$(var merger/output/objects)"/>
+      <arg name="priority_mode" value="1"/>
```
[the remark this answers is not in the document]
Comment (yukkysaito) [reply launch/tier4_perception_launch/launch/object_recognition/detection/lidar_based_detection.launch.xml:129]: Detection by tracker stabilize the shape size. For the same object, the stable result from Detection by tracker is preferred.

[the remark this answers is not in the document]
Comment (yukkysaito) [reply launch/tier4_perception_launch/launch/object_recognition/detection/lidar_based_detection.launch.xml:129]: memo : use confidence based priority mode

[the remark this answers is not in the document]
Comment (yukkysaito) [reply launch/tier4_perception_launch/launch/object_recognition/detection/lidar_based_detection.launch.xml]: https://github.com/autowarefoundation/autoware.universe/pull/1683/commits/b73e8eca83b38d71a5c73f3764b990a9554df2db

File: perception/shape_estimation/lib/corrector/bus_corrector.cpp [hunk 2]
```diff
@@ -17,8 +17,20 @@
...
+    return corrector_utils::correctVehicleBoundingBoxWithReferenceShape(
+      ref_shape_size_info_.get(), shape, pose);
+  } else if (use_reference_yaw_) {  // use reference yaw
+    return corrector_utils::correctVehicleBoundingBoxWithReferenceYaw(params_, shape, pose);
+  } else if (use_reference_shape_size) {  // use reference shape size
+    RCLCPP_WARN(
+      rclcpp::get_logger("shape_estimation"),
+      "Currently, not support reference yaw is false, reference size is true");
+    return corrector_utils::correctVehicleBoundingBox(
+      params_, shape, pose);  // TODO(yukkysaito) implement only reference shape size version
```
[the remark this answers is not in the document]
Comment (yukkysaito) [reply perception/shape_estimation/lib/corrector/bus_corrector.cpp]: fixed [d265936](https://github.com/autowarefoundation/autoware.universe/pull/1683/commits/d2659365748b5be6a35c5d215b05373e7e5ec9fb)

**Comment by yukkysaito** [comment]
@yukke42 Can I ask for a review?

**Comment by yukke42** [comment]
> @yukke42 Can I ask for a review?

Sorry for the late, I'll check in this week.

File: perception/shape_estimation/lib/corrector/bus_corrector.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (yukke42) [reply perception/shape_estimation/lib/corrector/bus_corrector.cpp]: `CarCorrector::correct`, `BusCorrector::correct` and `TruckCorrector::correct` look like the same implementation. You can move the implementation to `VehicleCorrector:correct` in vehicle_corrector.hpp and delete these functions.

[the remark this answers is not in the document]
Comment (yukkysaito) [reply perception/shape_estimation/lib/corrector/bus_corrector.cpp]: Thank you!
I already implement VehicleCorrector:correct in vehicle_corrector.hpp and remove from CMakeLists.txt.
https://github.com/autowarefoundation/autoware.universe/pull/1683/commits/d2659365748b5be6a35c5d215b05373e7e5ec9fb#diff-21a7a98fbfe0f6b054231b09dbbd04a8d95e5ff19397132a63722ca4e31c8db1L30-L32
But I forgot to remove these source files :pray:

[the remark this answers is not in the document]
Comment (yukkysaito) [reply perception/shape_estimation/lib/corrector/bus_corrector.cpp]: fixed at https://github.com/autowarefoundation/autoware.universe/pull/1683/commits/050d035eeab6afa92db26c7e6fd0489c054cd83d

**Comment by yukkysaito** [comment]
@miursh Can you review this PR?

File: launch/tier4_perception_launch/launch/object_recognition/detection/camera_lidar_fusion_based_detection.launch.xml [hunk 3]
```diff
@@ -229,8 +230,9 @@
     <let name="merger/output/objects" value="$(var output/objects)" unless="$(var use_object_filter)"/>
     <include file="$(find-pkg-share object_merger)/launch/object_association_merger.launch.xml">
       <arg name="input/object0" value="temporary_merged_objects"/>
-      <arg name="input/object1" value="clustering/objects"/>
+      <arg name="input/object1" value="detection_by_tracker/objects"/>
       <arg name="output/object" value="$(var merger/output/objects)"/>
+      <arg name="priority_mode" value="1"/>
```
Comment (miursh) [inline launch/tier4_perception_launch/launch/object_recognition/detection/camera_lidar_fusion_based_detection.launch.xml:235]: detection by tracker objects are prioritized only in camera_lidar_fusion.
Isn't is necessary to change also in lidar_based_detection.launch.xml?

**Review by miursh — COMMENTED** [review]
In this PR, the merge priority remains based on confidence, right?

File: launch/tier4_perception_launch/launch/object_recognition/detection/camera_lidar_fusion_based_detection.launch.xml (hunk 3 above)
Comment (yukkysaito) [reply launch/tier4_perception_launch/launch/object_recognition/detection/camera_lidar_fusion_based_detection.launch.xml:235]: https://github.com/autowarefoundation/autoware.universe/pull/1683#discussion_r972608374
I was making a mistake. I'll change to confidence based mode.

Comment (yukkysaito) [reply launch/tier4_perception_launch/launch/object_recognition/detection/camera_lidar_fusion_based_detection.launch.xml]: fixed at https://github.com/autowarefoundation/autoware.universe/pull/1683/commits/c7fe9e613c864d42cb8cb019e36c9aa8ee7fc6de

**Review by miursh — APPROVED** [review]
LGTM

**Pull request merged**